hangman/lib/hangman/game.ex

```diff
@@ -0,0 +1,175 @@
+defmodule GameState do
+  defstruct game_state: :initializing,
+            turns_left: 7,
+            letters: [],
+            used: [],
+            last_guess: ""
+end
+
+defmodule Hangman.Game do
+
+  # Create a new game state
+  def new_game() do
+    # Pick a random word from the dictionary
+    word = Dictionary.random_word()
+    |> String.codepoints()
+
+    # Initialize the game state utilizing the random word
+    %GameState{letters: _word} = struct(%GameState{}, letters: word)
+  end
+
+  # Just in case someone plays the non-wrapper version. Return the game status.
+  def tally( game = %GameState{ game_state: :won  } ), do: game
+  def tally( game = %GameState{ game_state: :loss } ), do: game
+
+  def tally(game) do
+    # "Massage" the display list of letters by listing "_" for each character not guessed
+    disp_letters = build_letters_state(game.game_state, game.letters, game.used)
+
+    # Output game state
+    %{
+      game_state: game.game_state,
+      turns_left: game.turns_left,
+      letters: disp_letters,
+      used: game.used,
+      last_guess: game.last_guess
+    }
+
+  end
+
+  # Display full word if game state is won or lost. Otherwise transform letters into current status
+  def build_letters_state(:won,  letters, _used), do: letters
+  def build_letters_state(:lost, letters, _used), do: letters
+  def build_letters_state(_,     letters,  used), do: transform_letters(letters, used)
+
+  # HELPER: Only shows letter in current board state if letter is found in "used"
+  def show_letter(true, letter),   do: letter
+  def show_letter(false, _letter), do: "_"
+
+  # Takes the current state of the board and displays letters only if found in "used" list.
+  def transform_letters([], _used), do: []
+  def transform_letters([ h | t ], used) do
+    [ (h in used) |> show_letter(h) | t |> transform_letters(used) ]
+  end
+
+  # Pass in a letter as current guess, and update game status as necessary.
+  @spec make_move(struct(), binary()) :: tuple()
+  def make_move(game, guess) do
+    updated_game = handle_guess(game, guess)
+    updated_tally = tally(updated_game)
+
+    {updated_game, updated_tally}
+  end
+
+  ##############################################################################
```

Comment:
There are two ways of handling this kind of chain of logic. The one you chose is to make each step call the subsequent step if it succeeds.

The other way is to have a top-level pipeline, where each step is self contained, and each uses information in the passed state set by previous states to decide what to do.

Which do you feel is easier to change?